sdk/src/types/block/address/restricted.rs

```diff
@@ -0,0 +1,323 @@
+        self.allowed_capabilities = allowed_capabilities.into();
+        self
+    }
+
+    /// Sets the allowed capabilities flags.
+    #[inline(always)]
+    pub fn set_allowed_capabilities(&mut self, allowed_capabilities: impl Into<AddressCapabilities>) -> &mut Self {
+        self.allowed_capabilities = allowed_capabilities.into();
+        self
+    }
+}
+
+impl TryFrom<Address> for RestrictedAddress {
+    type Error = Error;
+
+    fn try_from(value: Address) -> Result<Self, Self::Error> {
+        Self::new(value)
+    }
+}
+
+impl core::fmt::Display for RestrictedAddress {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "{}", prefix_hex::encode(self.pack_to_vec()))
+    }
+}
+
+/// All possible capabilities that an [`Address`] can have.
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+#[non_exhaustive]
+pub enum AddressCapabilityFlag {
+    /// Can receive Outputs with Native Tokens.
+    NativeTokens,
+    /// Can receive Outputs with Mana.
+    Mana,
+    /// Can receive Outputs with a Timelock Unlock Condition.
+    OutputsWithTimelock,
+    /// Can receive Outputs with an Expiration Unlock Condition.
+    OutputsWithExpiration,
+    /// Can receive Outputs with a Storage Deposit Return Unlock Condition.
+    OutputsWithStorageDeposit,
+    /// Can receive Account Outputs.
+    AccountOutputs,
+    /// Can receive NFT Outputs.
+    NftOutputs,
+    /// Can receive Delegation Outputs.
+    DelegationOutputs,
+}
+
+impl AddressCapabilityFlag {
+    const NATIVE_TOKENS: u8 = 0b00000001;
+    const MANA: u8 = 0b00000010;
+    const OUTPUTS_WITH_TIMELOCK: u8 = 0b00000100;
+    const OUTPUTS_WITH_EXPIRATION: u8 = 0b00001000;
+    const OUTPUTS_WITH_STORAGE_DEPOSIT: u8 = 0b00010000;
+    const ACCOUNT_OUTPUTS: u8 = 0b00100000;
+    const NFT_OUTPUTS: u8 = 0b01000000;
+    const DELEGATION_OUTPUTS: u8 = 0b10000000;
+
+    /// Converts the flag into the byte representation.
+    pub fn as_byte(&self) -> u8 {
+        match self {
+            Self::NativeTokens => Self::NATIVE_TOKENS,
+            Self::Mana => Self::MANA,
+            Self::OutputsWithTimelock => Self::OUTPUTS_WITH_TIMELOCK,
+            Self::OutputsWithExpiration => Self::OUTPUTS_WITH_EXPIRATION,
+            Self::OutputsWithStorageDeposit => Self::OUTPUTS_WITH_STORAGE_DEPOSIT,
+            Self::AccountOutputs => Self::ACCOUNT_OUTPUTS,
+            Self::NftOutputs => Self::NFT_OUTPUTS,
+            Self::DelegationOutputs => Self::DELEGATION_OUTPUTS,
+        }
+    }
+
+    /// Returns the index in [`AddressCapabilities`] to which this flag is applied.
+    pub fn index(&self) -> usize {
+        match self {
+            Self::NativeTokens
+            | Self::Mana
+            | Self::OutputsWithTimelock
+            | Self::OutputsWithExpiration
+            | Self::OutputsWithStorageDeposit
+            | Self::AccountOutputs
+            | Self::NftOutputs
+            | Self::DelegationOutputs => 0,
+        }
+    }
```

Comment:
Is this supposed to return the `Flag Index`? Is this correct atm?

Reply:
Yes it's correct, it's the byte index

Reply:
Yea theyre all index 0 atm, first 8. the next flag would be index 1